### DISCUSSION TITLE
Composer: private vcs repositories includes dist

### DISCUSSION THREAD

**Discussion by DjordyKoert** [renovatebot/renovate#20609] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.152.4

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

15.8.0 EE

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Wanted result in `composer.lock`:
```
{
            "name": "company/some-private-package",
            "version": "3.1.0",
            "source": {
                "type": "git",
                "url": "git@gitlab.company.com:packages/some-private-package.git",
                "reference": "036d7a077c47381221f74f080aca23553e264dd1"
            },
            ...
}
```

Actual result  in `composer.lock`:
```
{
            "name": "company/some-private-package",
            "version": "3.1.0",
            "source": {
                "type": "git",
                "url": "git@gitlab.company.com:packages/some-private-package.git",
                "reference": "036d7a077c47381221f74f080aca23553e264dd1"
            },
            "dist": {
                "type": "zip",
                "url": "https://gitlab.company.com/api/v4/projects/packages%2Fsome-private-package/repository/archive.zip?sha=036d7a077c47381221f74f080aca23553e264dd1",
                "reference": "036d7a077c47381221f74f080aca23553e264dd1",
                "shasum": ""
            },
            ...
}
```

This causes our composer install to install from dist which results in a 404 & install from source:
Command used: `composer install -o -a`
Results in: 
```
    Failed to download company/some-private-package from dist: The "https://gitlab.company.com/api/v4/projects/packages%2Fsome-private-package/repository/archive.zip?sha=036d7a077c47381221f74f080aca23553e264dd1" file could not be downloaded (HTTP/2 404 ):
{"message":"404 Project Not Found"}
    Now trying to download from source
  - Syncing company/some-private-package (3.1.0) into cache
```

### What you tried so far.

- Tried setting `gitUrl` to `"ssh"`
  - It now clones our initial repositories using ssh, instead of https (Which is good)
  - It did not fix the problem with `composer update` using https
- Tried debugging with ssh key
  - Checked if our private git repositories for our packages were cloneable with git (They are) using a local run of the docker image and using a `git clone`.

Our current fix is to manually checkout the branch for our private package & manually run a `composer update`, which results in the removal of the `"dist"` inside of the `composer.lock`.

### Relevant debug logs

<details><summary>Logs</summary>

## DEBUG: packageFiles with updates (repository=group/repository, baseBranch=master)
```
...
               {
                 "depType": "require",
                 "depName": "company/some-private-package",
                 "currentValue": "^2.4",
                 "datasource": "git-tags",
                 "packageName": "git@gitlab.company.com:packages/some-private-package.git",
                 "lockedVersion": "2.4.1",
                 "depIndex": 63,
                 "updates": [
                   {
                     "bucket": "major",
                     "newVersion": "3.1.0",
                     "newValue": "^3.0",
                     "newDigest": "6aa6b40d79e59e033f96a31a8ddcb298dab5d626",
                     "newMajor": 3,
                     "newMinor": 1,
                     "updateType": "major",
                     "isRange": true,
                     "branchName": "renovate/company-some-private-package-3.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "composer",
                 "currentVersion": "2.4.1",
                 "isSingleVersion": false,
                 "fixedVersion": "2.4.1"
               },
...
```

</details>

<details><summary>composer.json</summary>

```json
...
    "repositories": [
        {
            "name": "company/some-private-package",
            "type": "vcs",
            "url": "git@gitlab.company.com:packages/some-private-package.git"
        },
...
```
</details>

**Comment by rarkins** [comment]
Renovate does not update the lock file directly. If there are changes in that lock file then they're made by `composer` itself. If you look in the debug logs you can see the exact `composer` commands which were run for the branch.

**Reply by DjordyKoert** [reply]
The only two composer commands that are being ran are:

- `composer install --ignore-platform-req='ext-*' --ignore-platform-req='lib-*' --no-ansi --no-interaction --no-scripts --no-autoloader --no-plugins`

- `composer update company/some-private-package --with-dependencies --ignore-platform-req='ext-*' --ignore-platform-req='lib-*' --no-ansi --no-interaction --no-scripts --no-autoloader --no-plugins`

Running both of these manually result in `"dist"` being removed. Which makes me think something weird is going on inside renovate

**Reply by rarkins** [reply]
If you can find the relevant source code which is doing it, let us know

**Reply by DjordyKoert** [reply]
I finally found a way to reproduce it!

By including a [gitlab token](https://getcomposer.org/doc/articles/authentication-for-private-packages.md#gitlab-token) inside the config causes the dist to appear inside of `composer.lock`. Renovate apparently sets the composer config in [lib/modules/manager/composer/artifacts.ts#L60-L66](https://github.com/renovatebot/renovate/blob/e8a5437cd3bc9a937948e09c5b1357d88c040af4/lib/modules/manager/composer/artifacts.ts#L60-L66) which is the result of the `"dist"` being added.
```ts
      authJson['gitlab-token'] = authJson['gitlab-token'] ?? {};
      authJson['gitlab-token'][host] = gitlabHostRule.token;
      // https://getcomposer.org/doc/articles/authentication-for-private-packages.md#gitlab-token
      authJson['gitlab-domains'] = [
        host,
        ...(authJson['gitlab-domains'] ?? []),
      ];
```

**Reply by rarkins** [reply]
Nice work! Setting the token if present sounds like the right behavior for Renovate by default. Strange that it produces a change in lock file though.

Do you use authentication at all for Composer? If not, then you'd probably want an option which says "Don't add any authentication before running composer"?

**Reply by DjordyKoert** [reply]
We authenticate using ssh keys for git (which we put inside a custom build image made from the original image). Which means there is no reason for us to set our gitlab-token inside the composer config, since all authentication is already handled. 

An option to, as you said, "Don't add any authentication before running composer" seems like a nice solution for us.

**Reply by rarkins** [reply]
In that case I think we add an new option to https://docs.renovatebot.com/configuration-options/#postupdateoptions

Please create a feature request and reproduction repo. Can you create one on github.com which points to a package on gitlab.com? i.e. if you then add a gitlab.com token it causes the dist part to be added

**Reply by DjordyKoert** [reply]
PR has been made https://github.com/renovatebot/renovate/pull/20634

Reproduction repo: https://github.com/DjordyKoert/renovate-20609

**Reply by DjordyKoert** [reply]
@rarkins Could we not include this as a per `hostRule` option? This way users can easily exclude a specific hostRule auth from being set inside of the composer-auth.

Maybe an option like `excludeFromComposerAuth` with type `boolean`?

**Reply by DjordyKoert** [reply]
As an example this would result in:
```
hostRules: [
    {
        hostType: 'github',
        matchHost: 'api.github.com',
        token: 'ghp_github-token',
    },
    {
        hostType: 'gitlab',
        matchHost: 'gitlab.com',
        token: 'gitlab-token',
        excludeFromComposerAuth: true,
    }
],

**Reply by rarkins** [reply]
I don't hate this idea. Potentially it could be a new boolean option `artifactAuth` which we use for wider purposes. Initially the only use is setting it to `false` for composer, but maybe in future it could be a way to include auth for arbitrary hosts, for example

**Reply by DjordyKoert** [reply]
If this option would be used for a wider purpose in the future, then wouldn't it be better for users to be able to define specific managers which they want to exclude auth from (using an `array` like `['composer', 'npm']`)?

Then more managers could easily be added in the future without breaking backwards compatibility.

**Reply by rarkins** [reply]
So then you need possible values which indicate:
- Default/current behavior (`null`?)
- Use this for no artifact auth (`[]`?)
- Use it for specific artifact auth (e.g. npm and not composer) (`["npm"]`)
